### ISSUE TITLE
About the processing of "pd_indexing" in index.py

### ISSUE THREAD

**Issue by isletennos** [chomeyama/SiFiGAN#6]
I have two questions regarding the "pd_indexing" process.

Firstly, why is "torch.arange" used to obtain "idxP" and "idxF"? 
Wouldn't the sequence generated by "torch.arange" have a greater impact than the "dilations" that are added later with "torch.add"? 
Is there any disadvantage to using "torch.ones"?

Secondly, it seems that the current implementation does not take into account mini-batches. 
The line "maxP = -((torch.min(idxP) + batch_length))" appears to be computing the "maxP" for the entire mini-batch. 
This means that if there are data with significantly different F0 values within a mini-batch, it seems that only one data will be affected. 
Is this intentional?

Thank you.

**Comment by chomeyama** [comment]
Sorry for not replying sooner.

> Firstly, why is "torch.arange" used to obtain "idxP" and "idxF"?
> Wouldn't the sequence generated by "torch.arange" have a greater impact than the "dilations" that are added later with "torch.add"?
> Is there any disadvantage to using "torch.ones"?

The "torch.arange" is used to prepare the base indexes of the input feature map. Here the base indexes mean the indexes to which the center kernel is applied when the kernel size is three.
IdxP and idxF specify the index to which the left and right kernel is applied.
So subtracted/added "dilations" means how far away from the base indexes the left/right kernels are applied on the input map.
Therefore, I believe "torch.arange", which efficiently returns the base indexes, is suitable here.
If you have ideas for better implementation, welcome to share!

> Secondly, it seems that the current implementation does not take into account mini-batches.
> The line "maxP = -((torch.min(idxP) + batch_length))" appears to be computing the "maxP" for the entire mini-batch.
> This means that if there are data with significantly different F0 values within a mini-batch, it seems that only one data will be affected.
> Is this intentional?

Yes, you are right.
The current implementation could be more efficient, as you mentioned.
We can improve it by not using the "pad1d" function and instead directly specifying the indexes as the following pseudo code:

```
B, C, T = x.size()
batch_index = torch.arange(0, B, dtype=torch.long, device=x.device).reshape(B, 1, 1)
ch_index = torch.arange(0, C, dtype=torch.long, device=x.device).reshape(1, C, 1)
dilations = torch.clamp((d * dilation).long(), min=1)
  
# get past index (assume reflect padding)
idx_base = torch.arange(0, T, dtype=torch.long, device=x.device).reshape(1, 1, T)
idxP = (idx_base - dilations).abs() % T
idxP = (batch_index, ch_index, idxP)

# get future index (assume reflect padding)
idxF = idx_base + dilations
overflowed = idxF >= T
idxF[overflowed] = -(idxF[overflowed] % T)
idxF = (batch_index, ch_index, idxF)
```

**Comment by isletennos** [comment]
Thank you for your response.

The first question is about my understanding that the receptive field changes depending on the "dilations". 
The "dilations" are generated based on F0, so their values vary within a certain range. Let's assume they vary between 0 and 500. 
Regarding the sequence generated by torch.arange, let's say we generate 8,400 samples. In this case, the sequence generated by torch.arange will be [0,1,2,3,...,8399]. 
If we add the "dilations" to this sequence and then take the maximum value, wouldn't the "dilations" only have an impact on the latter part of the sequence? 
For example, if the first half of the "dilations" values were around 500 Hz, and the second half were around 100 Hz, wouldn't the influence of the first half of the "dilations" be ignored?
Please correct me if I'm wrong, but if my understanding is correct, would it be better to use "dilations" instead of idxP when calculating maxP and maxF?
ex)
`maxP = ((torch.max(dilations) + batch_length))`

**Comment by chomeyama** [comment]
> If we add the "dilations" to this sequence and then take the maximum value, wouldn't the "dilations" only have an impact on the latter part of the sequence?
> For example, if the first half of the "dilations" values were around 500 Hz, and the second half were around 100 Hz, wouldn't the influence of the first half of the "dilations" be ignored?

Yes, that's true.
If the first half of the "dilations" values were calculated on F0 at around 500 Hz, and those in the second half were around calculated on F0 at around 100 Hz, the maximum value (idxF) depends on only the second half.
As you know, it's because the first half (500 Hz) of "dilations" has smaller values than the second half (100 Hz).

About your proposal, I think the merit of using `torch.max(dilations)` is that we can reduce the number of times "torch.max" is called, which is now called twice at `maxP = -((torch.min(idxP) + batch_length))` and `maxF = torch.max(idxF) - (batch_length - 1)`, to one, right?
That's because `torch.max(dilations)` gives us the upper bound on the amount we should pad for both "maxP" and "maxF", right?

Please tell me if my understanding of your question/proposal is not correct!

**Comment by isletennos** [comment]
I apologize for my poor explanation.

My understanding is that the receptive field of PDCNN depends only on the "dilations" derived from F0.
If we take the MAX from the value obtained by adding "dilations" to the array generated by "torch.arange", then the receptive field would depend on both F0 and the output of "torch.arange".
Furthermore, in terms of value scaling, it would heavily depend on the output of "torch.arange", which would contradict the concept of the receptive field changing based on F0.

My proposal is to eliminate impurities and make the receptive field depend only on the "dilations" derived from F0.

**Comment by chomeyama** [comment]
> My understanding is that the receptive field of PDCNN depends only on the "dilations" derived from F0.

Yes, that's true!

> If we take the MAX from the value obtained by adding "dilations" to the array generated by "torch.arange", then the receptive field would depend on both F0 and the output of "torch.arange".

Well, I think the receptive field doesn't depend on the output of "torch.arange"...
The receptive field size of each time step has already been defined by "dilations" on line 28.
`dilations = d * dilation`

The "torch.min(idxP)" and "torch.max(idxF)" operations are called to get the required padding sizes defined by "dilations."
And the padding size doesn't affect the receptive field size, while the receptive field size affects the padding size, right?

However, there might still be differences in our understanding of our questions/answers...
If so, it would be helpful to show me a concrete code of your proposal!

**Comment by stealthinu** [comment]
I am modifying the code that uses SiFiGAN as a decoder in the open-source project MMVC, which is a real-time voice change tool. The author of this issue, isletennos, is also involved in the development of MMVC.

I hope you don't mind me joining the conversation.

The variable "d" in the function "bin.train.Trainer._train_step" is transformed as follows:
```python
    d = tuple([d[:1].to(self.device) for d in d])
```
As a result of this transformation, the size of the tensors in "d" changes. For example, if batch_size is 8, then the sizes of the tensors in "d" before and after the transformation are:

Before:
d[0].size() : torch.Size([8, 1, 512])
d[1].size() : torch.Size([8, 1, 2048])
d[2].size() : torch.Size([8, 1, 4096])
d[3].size() : torch.Size([8, 1, 8192])

After:
d[0].size() : torch.Size([1, 1, 512])
d[1].size() : torch.Size([1, 1, 2048])
d[2].size() : torch.Size([1, 1, 4096])
d[3].size() : torch.Size([1, 1, 8192])

In other words, only the information from the first batch is retained in "d".

In the function "utils.index.pd_indexing", "xP" and "xF" are generated from "x" and "d". However, while "x" uses information from all batches, "d" only uses information from the first batch.

I modified the code to use information from all batches to generate "xP" and "xF", and the modified code can be found at the following link:
https://github.com/stealthinu/SiFiGAN/tree/batch-aware-d

I would like to know if their modification is correct.

Thank you.

**Comment by chomeyama** [comment]
@stealthinu 
Thank you for your comment and explanation!

> As a result of this transformation, the size of the tensors in "d" changes.

Oh, this is my mistake...
Thank you for notifying me, I will fix it!

> I modified the code to use information from all batches to generate "xP" and "xF", and the modified code can be found at the following link:

Finally, I got it!
Yes, your modification looks reasonable to me.
Before your modification, I think the indexing operation is performed on unexpectedly shifted sequences, right?

**Comment by stealthinu** [comment]
Yes, I think so.
Actually, when debugging the modified code, I checked that different values were selected compared to the original code.

I intended to make minimal changes from the original code in this modification.
However, y-chan's suggested modification seems simpler and better.

**Comment by chomeyama** [comment]
> Actually, when debugging the modified code, I checked that different values were selected compared to the original code.

OK, I see. Thank you for teaching me the mistake!

> However, y-chan's suggested modification seems simpler and better.

Looks good!

**Comment by isletennos** [comment]
As for my concerns, their code has addressed them!
Thank you!

**Comment by chomeyama** [comment]
@isletennos 

OK! Thank you for notifying and discussing me!
I close this issue👍
